feat: implement kafka and redis

anticlown322 opened the pull request.

anticlown322 committed:
feat: start implementing kafka

anticlown322 committed:
feat + fix: implement kafka

anticlown322 committed:
feat + fix: add kafka to payouts + correct balance after bet

anticlown322 committed:
feat: implement redis

anticlown322 committed:
refactor: minor changes

DarkKezar commented on src/backend/BettingService.BLL/UseCases/Bets/Commands/PlaceBet/PlaceBetCommandHandler.cs:
```diff
@@ -2,73 +2,89 @@
-        }
+            BetId = betId,
+            CorrelationId = correlationId,
+            ValidationType = "User",
```
String "User" like magic number. Create abstraction with all possible types

DarkKezar commented on src/backend/BettingService.BLL/UseCases/Bets/Commands/PlaceBet/PlaceBetCommandHandler.cs:
```diff
@@ -2,73 +2,89 @@
+            CorrelationId = correlationId,
+            ValidationType = "User",
+            Username = request.Username,
+            Amount = (double)request.PlaceBetDto.Amount,
```
What is the reason of type casting?

DarkKezar commented on src/backend/BettingService.BLL/UseCases/Payouts/Commands/RequestPayout/RequestPayoutCommandHandler.cs:
```diff
@@ -42,30 +47,74 @@ public async Task<Unit> Handle(RequestPayoutCommand request, CancellationToken c
+        {
+            var message = await kafkaConsumer.ConsumeSingleMessageAsync<PayoutResult>(
+                kafkaSettings.Value.Topics.PayoutResults,
+                TimeSpan.FromSeconds(5),
```
Magic number

DarkKezar commented on src/backend/SportDataService.GrpcService/Services/Kafka/KafkaConsumerService.cs:
```diff
@@ -0,0 +1,35 @@
+using SportDataService.GrpcService.Models.Settings;
+using SportDataService.GrpcService.Utility;
+
+namespace SportDataService.GrpcService.Services.Kafka;
```
Stange namespace. Kafka in grpc namespace

DarkKezar commented on src/backend/SportDataService.Infrastructure/Repository/Cached/CachedMatchRepository.cs:
```diff
@@ -0,0 +1,121 @@
+
+namespace SportDataService.Infrastructure.Repository.Cached;
+
+public sealed class CachedMatchRepository(
```
When you declare difficult class with logic and fields - try avoid such constructors. You can control access modifiers of fields. Rewrite with full fields declaration and common constructor

DarkKezar commented on src/backend/SportDataService.Infrastructure/Repository/Cached/CachedTeamRepository.cs:
```diff
@@ -0,0 +1,99 @@
+
+namespace SportDataService.Infrastructure.Repository.Cached;
+
+public sealed class CachedTeamRepository(
```
You have several CachedRepository, use abstraction **_IChacheRepository<T>_** and base implementation **_BaseChacheRepository<T>:IChacheRepository<T>_**. Then, all your such classes will declared as **_TeamRepository:BaseChacheRepository<Team>, ITeamRepository_**, where **_ITeamRepository:IChacheRepository<Team>_**
That way you will get right tree of abstractions and DRY principle implementation.

DarkKezar commented on src/backend/compose.yaml:
docker-compose must be own for each micro service. You lost sense of micro services architecture now.
Possible questions:

How I should use each services without dependency of each other?
How we replicate one or several services with such docker-compose file?
In case, when we update any one service, why we should reboot others?

anticlown322 committed:
refactor: check description

- split docker-compose into multiple file
- change project structure
- add run.bat script
- remove unused asseblies

anticlown322 committed:
refactor: minor changes

anticlown322 committed:
refactor: change repositories hierarchy

anticlown322 committed:
refactor: move kafka from grpc to infrastructure

anticlown322 committed:
refactor: decompose data collection service
